### DISCUSSION TITLE
Running generator on swapi, getting compiler error

### DISCUSSION THREAD

**Discussion by bvankuik** [maticzav/swift-graphql#118] [Q&A] [answered]
I ran the generator on the swapi, as follows:

    $ swift-graphql -o Generated.swift https://swapi-graphql.netlify.app/.netlify/functions/index

I started a new project, and included the above generated file, and typed the following code:

    struct MyStarship {
        var name: String
    }

And:

    let url = URL(string: "https://swapi-graphql.netlify.app/.netlify/functions/index")!
    let client = SwiftGraphQLClient.Client(request: URLRequest(url: url))
    let starshipSelection = Selection.Starship {
        MyStarship(name: try $0.name())
    }
    let query = Objects.Root.starship(selection: starshipSelection))
    client.query(query)

I'm getting the following compiler error:

    Cannot convert value of type 'Selection<MyService.MyStarship?, Objects.Starship>'
    to expected argument type 'Selection<MyService.MyStarship?, Objects.Starship?>'

Looking at the generated code, it states the following extension:

    extension Selection where T == Never, TypeLock == Never {
        public typealias Starship<T> = Selection<T, Objects.Starship>
    }

However, it also states:

    extension Fields where TypeLock == Objects.Root {
        public func starship<T>(
        id: OptionalArgument<String> = .init(), starshipId: OptionalArgument<String> = .init(),
        selection: Selection<T, Objects.Starship?>
      ) throws -> T {
          // ...
    }

So I wrote out all lines with explicit types, to make sure I understood the error:

        let decoder: (Fields<Objects.Starship>) throws -> MyStarship = {
            MyStarship(name: try $0.name() ?? "")
        }
        let starshipSelection = Selection<MyStarship, Objects.Starship?>(decoder: decoder)
        let query = Objects.Root.starship(selection: starshipSelection))
        client.query(query)

The compiler complains at the `let starshipSelection`, saying:

    Cannot convert value of type '(Fields<Objects.Starship>) throws -> MyService.MyStarship'
    to expected argument type '(Fields<Objects.Starship?>) throws -> MyService.MyStarship'

I don't really understand what's going on here. Why must the `Objects.Root.starship(selection:)` parameter be of type `Selection<MyStarship, Objects.Starship?>` ? That'd basically mean that the `Fields` object is typed to an optional `Objects.Starship` but such an object doesn't make sense, IMHO.

I have the feeling that I'm overlooking something. What's going on?

**Comment by maticzav** [comment, marked as answer]
From the top of my head, I think you need to make your query nullable.

```swift
let query = Objects.Root.starship(selection: starshipSelection.nullable)
client.query(query)
```

If you check the error you are getting

> Cannot convert value of type 'Selection<MyService.MyStarship?, Objects.Starship>' to expected argument type 'Selection<MyService.MyStarship?, Objects.Starship?>'

It's saying that the TypeLock should be nullable but yours is not (the `Objects.Starship` vs `Objects.Starship?`).
